README.md
CCM is the Configuration Cache Manager

# Description

These modules handle the conversion of an XML or JSON profile into a
local binary cache, and give the API for Quattor modules to access
these caches.

If you are writing a Quattor-client module, all you probably need is
the `getElement` and `getTree` methods from a `Configuration`
object.  Typically you will combine them like this:

```perl
my $tree = $cfg->getElement("/foo/bar")->getTree();
```

And you will have a reference to a data structure, identical to what
you defined in your profile.

For more information, see the man pages.


ChangeLog
2012-01-25  Luis Fernando Muñoz Mejías

	* Release: 2.3.0
	- Fix the dependencies for Morgan-Stanley's changes.

	The new backend formats and the cleanups associated to them well
	deserve a new version!

2012-01-02  Christos Triantafyllidis,ctria@grid.auth.gr

	* Release: 2.2.9
	- LWP::UserAgent 6.x needs additional ssl options.

2009-09-02  Luis Fernando Muñoz Mejías

	* Release: 2.2.8
	- Relax the dependencies, as per bug #55179.

2009-03-04  Luis Fernando Muñoz Mejías <Luis.Fernando.Munoz.Mejias@cern.ch>

	* Release: 2.2.7
	- Convert REPORT and TABLE types to actual synonims of NLIST
	instead of subtypes.

2008-12-12  Marco Emilio Poleggi,<Marco.Poleggi@cnaf.infn.it>

	* Release: 2.2.6
	- [your comment here]

2008-12-12  Marco Emilio Poleggi,<Marco.Poleggi@cnaf.infn.it>

	* Release: 2.2.5
	- [your comment here]

2008-11-06  "Luis Fernando"

	* Release: 2.2.4
	- Fixing a warning related to the encoding of fetched information.

2008-11-05  "Luis Fernando"

	* Release: 2.2.3
	- Whitespace and pretty printing changes.

2008-11-05  "Luis Fernando"

	* Release: 2.2.2
	- Removing remaining methods referencing FETCH, LINK, STREAM and EMBED.

2008-11-04  "Luis Fernando"

	* Release: 2.2.1
	- Reduced the "base_url" is not defined message to debug level.
	
	During the workshop in Amsterdam we decided that this message caused
	unneeded worries and fear to users. Debug is OK for this, as the module
	works just fine without the base_url parameter.

2008-10-31  "Luis Fernando"

	* Release: 2.2.0
	- Removed support for types fetch, embed, stream and link.
	
	These types are no longer supported by the compiler, and thus during the
	workshop it was agreed to drop them from CCM too.

2008-07-18  "Luis Fernando"

	* Release: 2.1.0
	- Added new getUnescapedName and unescape.
	Let's stop copying and pasting this everywhere!

2008-07-14  "Luis Fernando"

	* Release: 2.0.8
	- Added an example /etc/ccm.conf to the RPM, and made it be installed as '.rpmnew' in case such file already exists.
	Should solve the documented (but forgotten by users) "failed to create Fetch object" on AII servers.

2008-07-07  "Luis Fernando"

	* Release: 2.0.7
	- Fixed the constructor of EDG::WP4::CCM::Configuration to get rid of taint warnings.

2008-05-28  root

	* Release: 2.0.6
	- Encode some values to UTF-8

2008-03-26  Luis Fernando Muñoz Mejías

	* Release: 2.0.5
	- Fixed the %post script trying to unconditionally kill ncm-cdispd, which
	could break installations.

2008-03-26  Luis Fernando Muñoz Mejías

	* Release: 2.0.4
	- Fixed bug #27529

2008-03-11  Luis Fernando Muñoz Mejías

	* Release: 2.0.3
	- Fixed a bug that had the potential to disable the entire notification system.
	Updated some tests.

2008-01-14  Rohitashva Sharma

	* Release: 2.0.2
	- improved fetching of foreign profiles
	- better singleton object for CacheManager

2007-11-27  Rohitashva Sharma

	* Release: 2.0.1
	- Solved bug#21170 (won't fetch profile if /etc/noquattor is present

2007-11-23  Rohitashva Sharma

	* Release: 2.0.0
	- Support for fetching foreign profiles 
	fix for bug#21846

2007-07-06  German Cancio Melia

	* Release: 1.5.15
	- Added failover URL to eg ccm.conf

2007-06-15  Charles Loomis

	* Release: 1.5.14
	- fix bug #27230; treat undef values as an empty string

2007-04-25  Marco Emilio Poleggi,<Marco.Poleggi@cern.ch>

	* Release: 1.5.13
	- ccm.conf: drop "profile_" prefix in XML filename

2007-03-22  Luis Fernando Muñoz Mejías

	* Release: 1.5.12
	- Removed getRecHash, following Germán's indications.

2007-03-21  Luis Fernando Muñoz Mejías

	* Release: 1.5.11
	- Added getTree method to Element class, as suggested in the workshop.

2007-02-26  Charles Loomis

	* Release: 1.5.10
	- use utf8 encoding to extract values from database

2007-02-26  Charles Loomis

	* Release: 1.5.9
	- use utf8 encoding for storing values in database

2007-02-26  Charles Loomis

	* Release: 1.5.8
	- allow UTF8 encoding of machine profiles

2007-02-07  German Cancio Melia

	* Release: 1.5.7
	- EDG::WP4::CCM::Element.pm: Added new getRecHash() function, Savannah #23574

2006-12-01  German Cancio Melia

	* Release: 1.5.6
	- Reduce number of hourly ccm-fetch invocations from 3 to 1 and increase time smearing to 1h

2006-11-25  German Cancio Melia

	* Release: 1.5.5
	- Now correctly handles the situation where force is defined but set to 0 in ccm-fetch - was causing to skip over if_modified_since() and thus downloading profile anyway

2006-11-23  German Cancio Melia

	* Release: 1.5.4
	- Fixed ccm-purge dying whenever finding a stale CCM active profile lock, now reporting as [INFO] and cleaning up

2006-11-23  German Cancio Melia

	* Release: 1.5.3
	- Fix Savannah #10558: Do skip over profile diff step when --force is set, in order to inconditionally update the CCM cache even if the new profile does not differ from the old one.

2006-11-07  German Cancio Melia

        * Release: 1.5.2
        - More verbose debug logging; bugfix to not loop over 304 (no chaHTTP profile status

2006-11-02  German Cancio Melia

        * Release: 1.5.1
        - Use standard Quattor-style timestamping when WARN/ERROR

2006-09-15  German Cancio Melia

        * Release: 1.5.0
        - Add support for a failover profile, in case the URL in --profile is not available. Useful in case of expired SSL certificates or broken main servers which may prevent profile updates to happen

2006-09-12  German Cancio Melia

        * Release: 1.4.33
        - Fix  #19416 - ccm-purge potentially removing active CCM profiles

2006-06-20  German Cancio Melia

        * Release: 1.4.32
        - Fix bug #15857 (CCM not correctly releasing lock files when instantiating multiple configurations)

2005-09-20  German Cancio Melia

        * Release: 1.4.31
        - Removed 'Provides' of edg-ccm as not liked (obsoletes/provides) by rpm 4.3.3

2005-09-20  German Cancio Melia

        * Release: 1.4.30
        - Changed cron job permissions to 0644, Savannah #10791

2005-08-18  Juan Antonio LOPEZ PEREZ


src/main/perl/CCfg.pm
# ${license-info}
# ${developer-info
# ${author-info}
# ${build-info}

package      EDG::WP4::CCM::CCfg;

use strict;
use LC::Exception qw(SUCCESS throw_error);
use Net::Domain qw(hostname hostdomain);

BEGIN{
 use      Exporter;
 use vars qw(@ISA @EXPORT @EXPORT_OK $VERSION);

 @ISA       = qw(Exporter);
 @EXPORT    = qw();           
 @EXPORT_OK = qw(initCfg getCfgValue);
 $VERSION   = sprintf("%d.%02d", q$Revision: 1.4 $ =~ /(\d+)\.(\d+)/);
}

=head1 NAME

EDG::WP4::CCM::CCfg

=head1 SYNOPSIS

  init()
    or 
  init("/etc/ccm.conf")

  $cache_root = getCfgValue ("cache_root");

=head1 DESCRIPTION

CCfg is used to get configuration parameters. Defualt values for
configuration parameters get overwritten if defined in configuration
file.

=over

=cut

# ------------------------------------------------------

my $ec = LC::Exception::Context->new->will_store_errors;
my $CONFIG_FN   = "ccm.conf";
my $DEF_EDG_LOC = "/usr";

my $cfg = {
	      "debug" => undef,
	      "force" => undef,
	      "profile" => undef,
	      "profile_failover" => undef,
	      "context" => undef,
	      "preprocessor" => undef,
	      "cache_root" => "/var/lib/ccm",
	      "get_timeout" => 30,
	      "lock_retries" => 3,
	      "lock_wait" => 30,
	      "retrieve_retries" => 3,
	      "retrieve_wait" => 30,
              "cert_file" => undef,
              "key_file" => undef,
              "ca_file" => undef,
              "ca_dir" => undef,
              "base_url" => undef,
              "world_readable" => undef,
              "dbformat" => "GDBM_File"
};

sub _resolveTags ($) {
  my ($s) = @_;
  if ($s=~/\$host/) {
    my $h = hostname();
    unless ($h) {
      throw_error ("could not resolve the hostname!");
      return();
    }
    $h = lc($h); # use lowercase for host.
    $s=~s/\$host/$h/g;
  }
  if ($s=~/\$domain/) {
    my $d = hostdomain();
    unless ($d) {
      throw_error ("could not resolve the domainname!");
      return();
    }
    $s=~s/\$domain/$d/g;
}
  return $s;
}

sub _readConfigFile ($) {
    my ($f) = @_;
    unless (open(CONF, "<$f")) {
	throw_error ("can't open config file: $f: $!");
	return ();
    }
    while (<CONF>) {
	if (/^\s*\#/) { next; }
	if (/^\s*$/) { next; }
	if (/^\s*(\w+)\s+(\S+)\s*$/) {
	    my $var = $1;
	    my $val = $2;
	    if ($var eq 'debug') {$cfg->{"debug"}=$val;}
	    elsif ($var eq 'force') {$cfg->{"force"}=$val;}
	    elsif ($var eq 'profile') {
		my $s = _resolveTags ($val);
		unless ($s) {
		    throw_error ("_resolveTags ($val)", $ec->error);
		    return ();  
		}
		$cfg->{"profile"}=$s;
	    }
	    elsif ($var eq 'profile_failover') {
		my $s = _resolveTags ($val);
		unless ($s) {
		    throw_error ("_resolveTags ($val)", $ec->error);
		    return ();
		}
		$cfg->{"profile_failover"}=$s;
	    }
	    elsif ($var eq 'context') {
		my $s = _resolveTags ($val);
		unless ($s) {
		    throw_error ("_resolveTags ($val)", $ec->error);
		    return ();  
		}
		$cfg->{"context"}=$s;
	    }
	    elsif ($var eq 'cache_root') {$cfg->{"cache_root"}=$val;}
	    elsif ($var eq 'get_timeout') {$cfg->{"get_timeout"}=$val;}
	    elsif ($var eq 'lock_retries') {$cfg->{"lock_retries"}=$val;}
	    elsif ($var eq 'lock_wait') {$cfg->{"lock_wait"}=$val;}
	    elsif ($var eq 'retrieve_retries') {$cfg->{"retrieve_retries"}=$val;}
	    elsif ($var eq 'retrieve_wait') {$cfg->{"retrieve_wait"}=$val;}
	    elsif ($var eq 'preprocessor') {$cfg->{"preprocessor"}=$val;}
            elsif ($var eq 'cert_file') {$cfg->{"cert_file"}=$val;}
            elsif ($var eq 'key_file') {$cfg->{"key_file"}=$val;}
            elsif ($var eq 'ca_file') {$cfg->{"ca_file"}=$val;}
            elsif ($var eq 'ca_dir') {$cfg->{"ca_dir"}=$val;}
            elsif ($var eq 'base_url') {$cfg->{"base_url"}=$val;}
            elsif ($var eq 'world_readable') {$cfg->{"world_readable"}=$val;}
            elsif ($var eq 'dbformat') {$cfg->{"dbformat"}=$val;}
            elsif ($var eq 'trust') {$cfg->{"trust"}=$val;}
	    else { throw_error("unknown config variable: $var"); }
	    next;
	}
	chomp;
	throw_error ("bad config file syntax: $_");
    }
    close(CONF);
    return SUCCESS;
}

=item initCfg (;$cfg_file)

Initialise CCfg. if $cfg_file parameter is present, file has to exists,
if it does not exist error is risen. If the parameter is not present
defualt EDG paths are used. If configuration file does not exist in defualt
locations the default values are used.

=cut

sub initCfg {
  my ($cp) = @_;
  if ($cp) {
    unless (-f $cp) {
      throw_error ("configuration file not found");
      return();
    } 
  } else {
    if (-f "/etc/$CONFIG_FN") {
      $cp = "/etc/$CONFIG_FN";
    } elsif (-f $DEF_EDG_LOC."/etc/$CONFIG_FN") {
      $cp = $DEF_EDG_LOC."/etc/$CONFIG_FN";
    } elsif (defined($ENV{"EDG_LOCATION"})
      && -f $ENV{"EDG_LOCATION"}."/etc/$CONFIG_FN") {
      $cp = $ENV{"EDG_LOCATION"}."/etc/$CONFIG_FN";
    } else {
      #no default configuration file exists
      #default parameters values will be used
      return();
    }
  }
  unless (_readConfigFile($cp)) {
    throw_error ("_readConfigFile($cp)", $ec->error);
    return ();
  }
  return SUCCESS;
}

=item getCfgValue ($key)

returns a value of the configuration parameter identified by $key.

=cut

sub getCfgValue ($) {
  my ($key) = @_;
  return ($cfg->{$key});
}

1;

__END__

=back

=head1 AUTHOR

Piotr Poznanski <Piotr.Poznanski@cern.ch>

=head1 VERSION

$Id: CCfg.pm.cin,v 1.4 2007/11/23 12:05:56 rsharma Exp $

=cut
